src/mame/ussr/juku.cpp

```diff
@@ -645,18 +639,35 @@ void juku_state::pio0_portc_w(uint8_t data)
 		floppy->ss_w(BIT(data, 6));
 	}
 
-	m_bank->set_bank(data & 0x03);
+	m_mode.select(data & 0b11);
 }
 
 void juku_state::machine_start()
 {
-	m_ram = std::make_unique<uint8_t []>(0x10000);
-
-	membank("rom_d800")->set_base(memregion("maincpu")->base() + 0x1800);
-
-	membank("ram_0000")->set_base(&m_ram[0x0000]);
-	membank("ram_4000")->set_base(&m_ram[0x4000]);
-	membank("ram_c000")->set_base(&m_ram[0xc000]);
+	// map memory modes
+	m_mode[0].install_rom(0x0000, 0x3fff, &m_rom[0x0000]);
+	m_mode[0].install_writeonly(0x0000, 0x3fff, &m_ram[0x0000]);
+	m_mode[0].install_ram(0x4000, 0xffff, &m_ram[0x4000]);
+	m_mode[1].install_ram(0x0000, 0xd7ff, &m_ram[0x0000]);
+	m_mode[1].install_rom(0xd800, 0xffff, &m_rom[0x1800]);
+	m_mode[1].install_writeonly(0xd800, 0xffff, &m_ram[0xd800]);
+	m_mode[2].install_ram(0x0000, 0x3fff, &m_ram[0x0000]);
+	// optional BASIC expansion cartridge
+	m_mode[2].install_rom(0x4000, 0xbfff, &m_exp[0x0000]);
+	//m_mode[2].install_ram(0x4000, 0xbfff, &m_ext[0][0x0000]);
+	// no info on programs actually using extra 32kb "memory window"
+	m_mode[2].install_writeonly(0x4000, 0xbfff, &m_ext[0][0x0000]);
+	// could also refer to usual ram (specs not clear)
+	//m_mode[2].install_writeonly(0x4000, 0xbfff, &m_ram[0x4000]);
+	m_mode[2].install_ram(0xc000, 0xd7ff, &m_ram[0xc000]);
+	m_mode[2].install_rom(0xd800, 0xffff, &m_rom[0x1800]);
+	m_mode[2].install_writeonly(0xd800, 0xffff, &m_ram[0xd800]);
+	// could as well disable the view, but for coherence
+	m_mode[3].install_ram(0x0000, 0xffff, &m_ram[0x0000]);
+	
+	// check if mouse is plugged
+	if (m_mouse.found() && strcmp(mconfig().options().mouse_device(), "none") != 0)
+		m_maincpu->space(AS_IO).install_read_handler(0x80, 0x80, read8smo_delegate(*m_mouse, FUNC(juku_mouse_device::mouse_port_r)));
```

Comment:
If you want to make the mouse optional, you need to turn it into a slot device.  See for example the Sharp X68000 mouse, or Sun workstation mouse.  The `mouse_device` option is for setting which host input devices should be enabled when the emulated system has mouse axis inputs.  You shouldn’t be using it like this.

Reply:
Technically nothing very bad happens if I just remove the mouse check and move the port allocation back to `io_map`. This will make the mouse always enabled, but on original system reading port 80 without a mouse attached stalled the system, so all the programs have to explicitly ask if the mouse is attached anyway. 

The authentic slot device would be a Multibus, which seems to be a lot of work that I would leave for next iterations.

Reply:
We do have at least some level of support for Multibus in src/devices/bus/multibus, so you hopefully won’t have to do everything from scratch for it.